external-crates/move/crates/move-regex-borrow-graph/src/regex.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt::{Debug, Display};
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) struct Regex<Lbl> {
+    labels: Vec<Lbl>,
+    ends_in_dot_star: bool,
+}
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) enum Extension<Lbl> {
+    Epsilon,
+    Label(Lbl),
+    DotStar,
+}
+
+//**************************************************************************************************
+// impls
+//**************************************************************************************************
+
+impl<Lbl> Regex<Lbl> {
+    pub fn epsilon() -> Self {
+        Self {
+            labels: vec![],
+            ends_in_dot_star: false,
+        }
+    }
+
+    pub fn label(lbl: Lbl) -> Self {
+        Self {
+            labels: vec![lbl],
+            ends_in_dot_star: false,
+        }
+    }
+
+    pub fn dot_star() -> Self {
+        Self {
+            labels: vec![],
+            ends_in_dot_star: true,
+        }
+    }
+
+    pub fn is_epsilon(&self) -> bool {
+        self.labels.is_empty() && !self.ends_in_dot_star
+    }
+
+    pub fn abstract_size(&self) -> usize {
+        1 + self.labels.len() + (self.ends_in_dot_star as usize)
+    }
+
+    /// Path for public facing API
+    pub fn pub_path(&self) -> (Vec<Lbl>, bool)
```

Comment:
By including this, you are preventing yourself from extending this to a different representation in the future.

Reply:
This is only an internal helper (note `Regex` has `pub(crate)`)